From 57f79f94ab0b6a662f1d4505ba653aa43447e02c Mon Sep 17 00:00:00 2001
From: Michal Kubecek <mkubecek@suse.cz>
Date: Mon, 11 Mar 2019 08:36:43 +0100
Subject: [PATCH 34/44] ethtool: set channel counts with SET_PARAMS request

Add support for setting channel counts using SET_PARAMS request with
ETHTOOL_A_PARAMS_CHANNELS nested attribute. This is a replacement for
ETHTOOL_SCHANNELS ioctl command.

Signed-off-by: Michal Kubecek <mkubecek@suse.cz>
---
 Documentation/networking/ethtool-netlink.txt | 16 +++--
 net/ethtool/common.c                         | 65 ++++++++++++++++++++
 net/ethtool/common.h                         |  3 +
 net/ethtool/ioctl.c                          | 59 +-----------------
 net/ethtool/params.c                         | 60 +++++++++++++++++-
 5 files changed, 139 insertions(+), 64 deletions(-)

diff --git a/Documentation/networking/ethtool-netlink.txt b/Documentation/networking/ethtool-netlink.txt
index ad21116016c3..b66b8870232b 100644
--- a/Documentation/networking/ethtool-netlink.txt
+++ b/Documentation/networking/ethtool-netlink.txt
@@ -484,10 +484,11 @@ SET_PARAMS
 ----------
 
 SET_PARAMS request modifies the settings retrieved by GET_PARAMS, i.e. it
-replaces ETHTOOL_SCOALESCE, ETHTOOL_SRINGPARAM and ETHTOOL_SPAUSEPARAM ioctl
-commands. For each of these, relevant data attributes are contained in
-a corresponding nest attribute. Some of the attributes provided by
-GET_SETPARAMS are read only and cannot be set by SET_PARAMS request.
+replaces ETHTOOL_SCOALESCE, ETHTOOL_SRINGPARAM, ETHTOOL_SPAUSEPARAM and
+ETHTOOL_SCHANNELS ioctl commands. For each of these, relevant data attributes
+are contained in a corresponding nest attribute. Some of the attributes
+provided by GET_SETPARAMS are read only and cannot be set by SET_PARAMS
+request.
 
     ETHTOOL_A_PARAMS_COALESCE		(nested)	coalescing parameters
         ETHTOOL_A_COALESCE_RX_USECS	    (u32)
@@ -521,6 +522,11 @@ GET_SETPARAMS are read only and cannot be set by SET_PARAMS request.
         ETHTOOL_A_PAUSE_AUTONEG		    (bool)
         ETHTOOL_A_PAUSE_RX		    (bool)
         ETHTOOL_A_PAUSE_TX		    (bool)
+    ETHTOOL_A_PARAMS_CHANNELS		(nested)	channel counts
+        ETHTOOL_A_CHANNELS_RX_COUNT	    (u32)
+        ETHTOOL_A_CHANNELS_TX_COUNT	    (u32)
+        ETHTOOL_A_CHANNELS_OTHER_COUNT	    (u32)
+        ETHTOOL_A_CHANNELS_COMBINED_COUN    (u32)
 
 
 Request translation
@@ -591,7 +597,7 @@ ETHTOOL_SRXFHINDIR		n/a
 ETHTOOL_GFEATURES		ETHNL_CMD_GET_SETTINGS
 ETHTOOL_SFEATURES		ETHNL_CMD_SET_SETTINGS
 ETHTOOL_GCHANNELS		ETHNL_CMD_GET_PARAMS
-ETHTOOL_SCHANNELS		n/a
+ETHTOOL_SCHANNELS		ETHNL_CMD_SET_PARAMS
 ETHTOOL_SET_DUMP		n/a
 ETHTOOL_GET_DUMP_FLAG		n/a
 ETHTOOL_GET_DUMP_DATA		n/a
diff --git a/net/ethtool/common.c b/net/ethtool/common.c
index 1ef13ec78c45..191198e7e671 100644
--- a/net/ethtool/common.c
+++ b/net/ethtool/common.c
@@ -4,6 +4,7 @@
 #include <linux/phy.h>
 #include <linux/net_tstamp.h>
 #include <net/devlink.h>
+#include <net/xdp_sock.h>
 #include "common.h"
 
 const char netdev_features_strings[NETDEV_FEATURE_COUNT][ETH_GSTRING_LEN] = {
@@ -224,3 +225,67 @@ int __ethtool_get_wol(struct net_device *dev, struct ethtool_wolinfo *wol)
 
 	return 0;
 }
+
+static int ethtool_get_max_rxfh_channel(struct net_device *dev, u32 *max)
+{
+	u32 dev_size, current_max = 0;
+	u32 *indir;
+	int ret;
+
+	if (!dev->ethtool_ops->get_rxfh_indir_size ||
+	    !dev->ethtool_ops->get_rxfh)
+		return -EOPNOTSUPP;
+	dev_size = dev->ethtool_ops->get_rxfh_indir_size(dev);
+	if (dev_size == 0)
+		return -EOPNOTSUPP;
+
+	indir = kcalloc(dev_size, sizeof(indir[0]), GFP_USER);
+	if (!indir)
+		return -ENOMEM;
+
+	ret = dev->ethtool_ops->get_rxfh(dev, indir, NULL, NULL);
+	if (ret)
+		goto out;
+
+	while (dev_size--)
+		current_max = max(current_max, indir[dev_size]);
+
+	*max = current_max;
+
+out:
+	kfree(indir);
+	return ret;
+}
+
+int __ethtool_set_channels(struct net_device *dev,
+			   const struct ethtool_channels *curr,
+			   struct ethtool_channels *channels)
+{
+	u16 from_channel, to_channel;
+	u32 max_rx_in_use = 0;
+	unsigned int i;
+
+	/* ensure new counts are within the maximums */
+	if (channels->rx_count > curr->max_rx ||
+	    channels->tx_count > curr->max_tx ||
+	    channels->combined_count > curr->max_combined ||
+	    channels->other_count > curr->max_other)
+		return -EINVAL;
+
+	/* ensure the new Rx count fits within the configured Rx flow
+	 * indirection table settings */
+	if (netif_is_rxfh_configured(dev) &&
+	    !ethtool_get_max_rxfh_channel(dev, &max_rx_in_use) &&
+	    (channels->combined_count + channels->rx_count) <= max_rx_in_use)
+	    return -EINVAL;
+
+	/* Disabling channels, query zero-copy AF_XDP sockets */
+	from_channel = channels->combined_count +
+		min(channels->rx_count, channels->tx_count);
+	to_channel = curr->combined_count + max(curr->rx_count, curr->tx_count);
+	for (i = from_channel; i < to_channel; i++)
+		if (xdp_get_umem_from_qid(dev, i))
+			return -EINVAL;
+
+	return dev->ethtool_ops->set_channels(dev, channels);
+}
diff --git a/net/ethtool/common.h b/net/ethtool/common.h
index 1e00a7eecbbd..5b05f1417b9e 100644
--- a/net/ethtool/common.h
+++ b/net/ethtool/common.h
@@ -21,6 +21,9 @@ int __ethtool_get_drvinfo(struct net_device *dev, struct ethtool_drvinfo *info);
 int __ethtool_get_ts_info(struct net_device *dev, struct ethtool_ts_info *info);
 int __ethtool_get_link(struct net_device *dev);
 int __ethtool_get_wol(struct net_device *dev, struct ethtool_wolinfo *wol);
+int __ethtool_set_channels(struct net_device *dev,
+			   const struct ethtool_channels *curr,
+			   struct ethtool_channels *channels);
 
 bool convert_legacy_settings_to_link_ksettings(
 	struct ethtool_link_ksettings *link_ksettings,
diff --git a/net/ethtool/ioctl.c b/net/ethtool/ioctl.c
index 484cde939527..3d04a11f1328 100644
--- a/net/ethtool/ioctl.c
+++ b/net/ethtool/ioctl.c
@@ -28,7 +28,6 @@
 #include <linux/sched/signal.h>
 #include <linux/net.h>
 #include <net/devlink.h>
-#include <net/xdp_sock.h>
 #include <net/flow_offload.h>
 #include <linux/ethtool_netlink.h>
 
@@ -845,37 +844,6 @@ void netdev_rss_key_fill(void *buffer, size_t len)
 }
 EXPORT_SYMBOL(netdev_rss_key_fill);
 
-static int ethtool_get_max_rxfh_channel(struct net_device *dev, u32 *max)
-{
-	u32 dev_size, current_max = 0;
-	u32 *indir;
-	int ret;
-
-	if (!dev->ethtool_ops->get_rxfh_indir_size ||
-	    !dev->ethtool_ops->get_rxfh)
-		return -EOPNOTSUPP;
-	dev_size = dev->ethtool_ops->get_rxfh_indir_size(dev);
-	if (dev_size == 0)
-		return -EOPNOTSUPP;
-
-	indir = kcalloc(dev_size, sizeof(indir[0]), GFP_USER);
-	if (!indir)
-		return -ENOMEM;
-
-	ret = dev->ethtool_ops->get_rxfh(dev, indir, NULL, NULL);
-	if (ret)
-		goto out;
-
-	while (dev_size--)
-		current_max = max(current_max, indir[dev_size]);
-
-	*max = current_max;
-
-out:
-	kfree(indir);
-	return ret;
-}
-
 static noinline_for_stack int ethtool_get_rxfh_indir(struct net_device *dev,
 						     void __user *useraddr)
 {
@@ -1534,9 +1502,6 @@ static noinline_for_stack int ethtool_set_channels(struct net_device *dev,
 						   void __user *useraddr)
 {
 	struct ethtool_channels channels, curr = { .cmd = ETHTOOL_GCHANNELS };
-	u16 from_channel, to_channel;
-	u32 max_rx_in_use = 0;
-	unsigned int i;
 	int ret;
 
 	if (!dev->ethtool_ops->set_channels || !dev->ethtool_ops->get_channels)
@@ -1547,29 +1512,7 @@ static noinline_for_stack int ethtool_set_channels(struct net_device *dev,
 
 	dev->ethtool_ops->get_channels(dev, &curr);
 
-	/* ensure new counts are within the maximums */
-	if (channels.rx_count > curr.max_rx ||
-	    channels.tx_count > curr.max_tx ||
-	    channels.combined_count > curr.max_combined ||
-	    channels.other_count > curr.max_other)
-		return -EINVAL;
-
-	/* ensure the new Rx count fits within the configured Rx flow
-	 * indirection table settings */
-	if (netif_is_rxfh_configured(dev) &&
-	    !ethtool_get_max_rxfh_channel(dev, &max_rx_in_use) &&
-	    (channels.combined_count + channels.rx_count) <= max_rx_in_use)
-	    return -EINVAL;
-
-	/* Disabling channels, query zero-copy AF_XDP sockets */
-	from_channel = channels.combined_count +
-		min(channels.rx_count, channels.tx_count);
-	to_channel = curr.combined_count + max(curr.rx_count, curr.tx_count);
-	for (i = from_channel; i < to_channel; i++)
-		if (xdp_get_umem_from_qid(dev, i))
-			return -EINVAL;
-
-	ret = dev->ethtool_ops->set_channels(dev, &channels);
+	ret = __ethtool_set_channels(dev, &curr, &channels);
 	if (ret == 0)
 		ethtool_notify(dev, NULL, ETHNL_CMD_SET_PARAMS,
 			       ETH_PARAMS_IM_CHANNELS, NULL);
diff --git a/net/ethtool/params.c b/net/ethtool/params.c
index 6381fb0a634c..73dc2111c32a 100644
--- a/net/ethtool/params.c
+++ b/net/ethtool/params.c
@@ -1,6 +1,7 @@
 /* SPDX-License-Identifier: GPL-2.0 WITH Linux-syscall-note */
 
 #include "netlink.h"
+#include "common.h"
 
 static const struct nla_policy get_params_policy[ETHTOOL_A_PARAMS_MAX + 1] = {
 	[ETHTOOL_A_PARAMS_UNSPEC]		= { .type = NLA_REJECT },
@@ -353,7 +354,7 @@ static const struct nla_policy set_params_policy[ETHTOOL_A_PARAMS_MAX + 1] = {
 	[ETHTOOL_A_PARAMS_COALESCE]		= { .type = NLA_NESTED },
 	[ETHTOOL_A_PARAMS_RING]			= { .type = NLA_NESTED },
 	[ETHTOOL_A_PARAMS_PAUSE]		= { .type = NLA_NESTED },
-	[ETHTOOL_A_PARAMS_CHANNELS]		= { .type = NLA_REJECT },
+	[ETHTOOL_A_PARAMS_CHANNELS]		= { .type = NLA_NESTED },
 };
 
 static const struct nla_policy coalesce_policy[ETHTOOL_A_COALESCE_MAX + 1] = {
@@ -574,6 +575,58 @@ static int update_pause(struct genl_info *info, struct net_device *dev,
 	return (ret < 0) ? ret : 1;
 }
 
+static const struct nla_policy channels_policy[ETHTOOL_A_CHANNELS_MAX + 1] = {
+	[ETHTOOL_A_CHANNELS_UNSPEC]		= { .type = NLA_REJECT },
+	[ETHTOOL_A_CHANNELS_MAX_RX]		= { .type = NLA_REJECT },
+	[ETHTOOL_A_CHANNELS_MAX_TX]		= { .type = NLA_REJECT },
+	[ETHTOOL_A_CHANNELS_MAX_OTHER]		= { .type = NLA_REJECT },
+	[ETHTOOL_A_CHANNELS_MAX_COMBINED]	= { .type = NLA_REJECT },
+	[ETHTOOL_A_CHANNELS_RX_COUNT]		= { .type = NLA_U32 },
+	[ETHTOOL_A_CHANNELS_TX_COUNT]		= { .type = NLA_U32 },
+	[ETHTOOL_A_CHANNELS_OTHER_COUNT]	= { .type = NLA_U32 },
+	[ETHTOOL_A_CHANNELS_COMBINED_COUNT]	= { .type = NLA_U32 },
+};
+
+static int update_channels(struct genl_info *info, struct net_device *dev,
+			   struct nlattr *nest)
+{
+	struct ethtool_channels old = { .cmd = ETHTOOL_GCHANNELS };
+	struct ethtool_channels new = { .cmd = ETHTOOL_SCHANNELS };
+	struct nlattr *tb[ETHTOOL_A_CHANNELS_MAX + 1];
+	bool mod = false;
+	int ret;
+
+	if (!nest)
+		return 0;
+	if (!dev->ethtool_ops->get_channels ||
+	    !dev->ethtool_ops->set_channels)
+		return -EOPNOTSUPP;
+	dev->ethtool_ops->get_channels(dev, &old);
+	new = old;
+	new.cmd = ETHTOOL_SCHANNELS;
+
+	ret = nla_parse_nested_strict(tb, ETHTOOL_A_CHANNELS_MAX, nest,
+				      channels_policy, info->extack);
+	if (ret < 0)
+		return ret;
+
+	if (ethnl_update_u32(&new.rx_count, tb[ETHTOOL_A_CHANNELS_RX_COUNT]))
+		mod = true;
+	if (ethnl_update_u32(&new.tx_count, tb[ETHTOOL_A_CHANNELS_TX_COUNT]))
+		mod = true;
+	if (ethnl_update_u32(&new.other_count,
+			     tb[ETHTOOL_A_CHANNELS_OTHER_COUNT]))
+		mod = true;
+	if (ethnl_update_u32(&new.combined_count,
+			     tb[ETHTOOL_A_CHANNELS_COMBINED_COUNT]))
+		mod = true;
+	if (!mod)
+		return 0;
+
+	ret = __ethtool_set_channels(dev, &old, &new);
+	return (ret < 0) ? ret : 1;
+}
+
 int ethnl_set_params(struct sk_buff *skb, struct genl_info *info)
 {
 	struct nlattr *tb[ETHTOOL_A_PARAMS_MAX + 1];
@@ -608,6 +661,11 @@ int ethnl_set_params(struct sk_buff *skb, struct genl_info *info)
 		goto out_ops;
 	if (ret)
 		req_mask |= ETH_PARAMS_IM_PAUSE;
+	ret = update_channels(info, dev, tb[ETHTOOL_A_PARAMS_CHANNELS]);
+	if (ret < 0)
+		goto out_ops;
+	if (ret)
+		req_mask |= ETH_PARAMS_IM_CHANNELS;
 
 	ret = 0;
 out_ops:
-- 
2.21.0

